### PR TITLE
feat(AutoscalePool): add support for max tasks per minute

### LINKED ISSUE #66: Autoscaling should enable max requests per minute setting
**Describe the feature**
Ability to set up `maxRequestsPerMinute` or similar option in the `autoscaledPoolOptions`.

**Motivation**
Concurrency cannot really be used to control how many requests will hit the target website because it depends on if the request takes 0.1 sec or 10 sec. This is a nice feature to handle blocking or not overloading target website.

**Constraints**
How it will work with other limits, I guess it will scale down either on concurrency or RPM, just depending on what comes first. But not sure how we would include it somehow to the metrics, etc.

cc @petrpatek if you have more ideas

### PR DESCRIPTION
**Pull request by vladfrangu**
Closes #66

### CODE REVIEW

**Comment by B4nan** [comment]
cc @mnmkng @pocesar

**Review by B4nan — APPROVED** [review]
lgtm

**Comment by vladfrangu** [comment]
From the original issue it seems to be a "no matter the concurrency, stop if it reaches this many tasks in that minute" (at least that's how I understood it and implemented it), but it's possible I could be wrong

**Comment by mnmkng** [comment]
Yeah, the original issue is fine and you're correct. It's just the counting of the minute that's incorrect IMO. You reset an interval once a minute and then add all the requests inside this interval. But it should not have an interval, it needs to track the time of tasks and then look back 60 seconds and count how many tasks.

An example, let's say the limit is 50.

We start the interval, nothing happens for 55 seconds, then the user starts 50 tasks. 5 secs later the limit kicks in, resets the counter and the user can immediately start 50 tasks again, running 100 tasks in 6 seconds.

The correct way to approach it is to:

1. when a user is starting a task, note the time T
2. find all the tasks which were started between T and T-60
3. count them, if the count is over limit, prevent starting task

Does that make sense?

**Comment by vladfrangu** [comment]
@mnmkng I think I moved it to a rolling interval, can you take a peek? 🙏

**Review by mnmkng — APPROVED** [review]

File: packages/core/src/autoscaling/autoscaled_pool.ts [hunk 1]
```diff
@@ -677,9 +677,7 @@ export class AutoscaledPool {
     protected async _incrementTasksDonePerSecond(intervalCallback: () => void) {
         this._tasksPerMinute.unshift(0);
 
-        if (this._tasksPerMinute.length > 60) {
-            this._tasksPerMinute.pop();
-        }
+        this._tasksPerMinute.pop();
```
Comment (mnmkng) [inline packages/core/src/autoscaling/autoscaled_pool.ts:680]: Haha, I had a comment half written that the if was useless, but in the end I was like "it doesn't hurt anything" 😄

File: packages/core/src/autoscaling/autoscaled_pool.ts [hunk 2]
```diff
@@ -640,6 +670,19 @@ export class AutoscaledPool {
 
         betterClearInterval(this.autoscaleInterval);
         betterClearInterval(this.maybeRunInterval);
+        if (this.tasksDonePerSecondInterval) betterClearInterval(this.tasksDonePerSecondInterval);
         if (this.snapshotter) await this.snapshotter.stop();
     }
+
+    protected async _incrementTasksDonePerSecond(intervalCallback: () => void) {
```
Comment (pocesar) [inline packages/core/src/autoscaling/autoscaled_pool.ts:677]: seems there's no need for this to be `async`, even the `intervalCallback` is synchronous 🙂

**Comment by pocesar** [comment]
I'd just point out that intervals are "best effort" and can be stuck if the event loop is busy, and can happen quite frequently on lower memory settings on the platform (extreme example: 0.03125 vCPU and 128MB, but seen happening on 1024MB as well), as they will be fighting: io, synchronous code and scheduled work, only one will win (hint, it's not the timers 😄)

I'm still sticking to my guns and trying to incentivize the rework on SDK 3 to have a better approach instead of porting the current way to the next major version

**Review by metalwarrior665 — COMMENTED** [review]
Thanks :)

**Comment by mnmkng** [comment]
Yeah, as Paulo says, doing this without intervals would be better, but this is good enough. We'll improve it if needed. It doesn't matter that much if the timer gets stuck for a few ms and lets one or two more tasks run.

**Comment by pocesar** [comment]
@mnmkng you couldn't be more wrong, sorry. even locally I can make the timers and console.log calls hang for 1-2 minutes with simple parsing of a big sitemap

**Comment by mnmkng** [comment]
@pocesar when you parse a sitemap for 2 mins, it doesn't matter too much if the timers get stuck, because everything gets stuck. As I said before, you're right. But in this case the real world impact of the stuck timers will be negligible.

**Comment by B4nan** [comment]
We agreed with @vladfrangu we will merge this as is, and we will be happy to revisit if @pocesar (or anybody else) can provide some way to break it. We want to have some initial beta published on NPM within next few weeks, so I hope it will be easier to reason about things like this after we do some real world testing.

**Pull request merged**